src/main/java/org/candlepin/controller/ConsumerMigration.java

```diff
@@ -0,0 +1,118 @@
+import org.slf4j.LoggerFactory;
+import org.xnap.commons.i18n.I18n;
+
+import java.util.List;
+import java.util.Objects;
+
+import javax.inject.Inject;
+
+/**
+ * Class responsible for migration of consumers from anonymous orgs.
+ */
+public class ConsumerMigration {
+
+    private static final Logger log = LoggerFactory.getLogger(ConsumerMigration.class);
+
+    private final OwnerCurator ownerCurator;
+    private final ConsumerCurator consumerCurator;
+    private final IdentityCertificateCurator identityCertificateCurator;
+    private final ContentAccessCertificateCurator contentAccessCertificateCurator;
+    private final CertificateSerialCurator serialCurator;
+    private final I18n i18n;
+    private final int batchSize;
+
+    @Inject
+    public ConsumerMigration(OwnerCurator ownerCurator, ConsumerCurator consumerCurator,
+        IdentityCertificateCurator identityCertificateCurator,
+        ContentAccessCertificateCurator contentAccessCertificateCurator,
+        CertificateSerialCurator serialCurator, I18n i18n, Configuration config) {
+        this.ownerCurator = Objects.requireNonNull(ownerCurator);
+        this.consumerCurator = Objects.requireNonNull(consumerCurator);
+        this.identityCertificateCurator = Objects.requireNonNull(identityCertificateCurator);
+        this.contentAccessCertificateCurator = Objects.requireNonNull(contentAccessCertificateCurator);
+        this.serialCurator = Objects.requireNonNull(serialCurator);
+        this.i18n = Objects.requireNonNull(i18n);
+        this.batchSize = config.getInt(ConfigProperties.CONSUMER_MIGRATION_BATCH_SIZE);
+    }
+
+    /**
+     * Migrates all consumers of the given anonymous org to the target org.
+     *
+     * @param originOwnerKey key of the source anonymous org
+     * @param destinationOwnerKey key of the destination org
+     */
+    public void migrate(String originOwnerKey, String destinationOwnerKey) {
+        Owner originOwner = this.ownerCurator.getByKey(originOwnerKey);
+        Owner destinationOwner = this.ownerCurator.getByKey(destinationOwnerKey);
+
+        List<String> consumerIds = this.ownerCurator.getConsumerIds(originOwner).list();
+        boolean failed = false;
+        for (List<String> consumerIdsBlock : Iterables.partition(consumerIds, this.batchSize)) {
+            try {
+                this.ownerCurator.transactional(args -> {
+                    migrateBatch(consumerIdsBlock, destinationOwner);
+                    return null;
+                }).execute();
+            }
+            catch (Exception e) {
+                log.warn("Failed to migrate a batch of consumers.", e);
+                failed = true;
+            }
+        }
+
+        if (failed) {
+            throw new ConsumerMigrationFailedException(i18n.tr("Failed to migrate consumers"));
+        }
+    }
+
+    private void migrateBatch(List<String> consumerIds, Owner destination) {
+        this.consumerCurator.lockAndLoad(consumerIds);
```

Comment:
Something I missed here is that this method locks the consumers **and loads the entire entities** into memory. which is something I was hoping to avoid by doing all the id-based queries below. I can't currently find in our curators an implementation for locking by ids, and only fetching the ids (instead of the entire entities), but I think its worth investigating if we can do it (it should be possible). E.g. by adding a method like `lockAndLoadIds`. If it is not possible to do this, it may not necessarily be bad if the batch size is kept small enough